### DISCUSSION TITLE
Initial conditions for use with the Jacobian debugger

### DISCUSSION THREAD

**Discussion by Sina-av** [idaholab/moose#25394] [Q&A Tools] [answered]
Hi all!

I want to check the Jacobian of a Moose app that I am writing with the help of the Jacobian debugger file, `analyze_jacobian.py`. For the simulations, I am using the following variables:

```
[Variables]
  [disp_x] []
  [disp_y] []
  [disp_z] []
  [NLvar] []
[]
```

Where `disp_x`, `disp_y`, and `disp_z` are the displacement degrees of freedom, and `NLvar` is just another variable governed by a PDE related to plastic deformation of the body.

Based on the webpage for the [Jacobian debugger](https://mooseframework.inl.gov/help/development/analyze_jacobian.html), it is recommended to use random initial conditions (ICs) to ensure that the values and gradients of the variables are non-zero.

In line with the use of ICs for the Jacobian debugger, I have some questions:

1) To clarify, is the Jacobian debugger only used for the IC and does not advance in time? Is there a way to run the Jacobian debugger for future time steps of a simulation?

2) If we can only run the Jacobian debugger for ICs of a system, I was wondering if there is a way to ensure plasticity in the model. In my Jacobian, I have derivatives of plastic deformation with respect to total deformation, and I want to check that as well. So, I was wondering if there is a way to ensure that, based on the given ICs, I will have plasticity in the system.

Thank you so much.

**Comment by lindsayad** [comment, marked as answer]
I myself do not use the `analyze_jacobian.py` script. I typically reduce my mesh size to be quite small, and then I run my simulation as normal, but with the PETSc command line options `-snes_test_jacobian -snes_test_jacobian_view` (these options are at the core of the `analyze_jacobian.py` script). With these options you can run your simulation for arbitrarily many time steps and so you will certainly test things like non-zero gradients.

I usually pair this with [DOFMap](https://mooseframework.inl.gov/source/outputs/DOFMapOutput.html) output to help ascertain what variable's Jacobian entries are bad

**Reply by Sina-av** [reply]
Great, thank you so much. This exactly did what I wanted.
